src/Nethermind/Nethermind.Blockchain/TransactionPools/TransactionPool.cs

```diff
@@ -0,0 +1,218 @@
+            IEthereumSigner signer, ILogManager logManager, int peerThreshold = 20)
+        {
+            _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+            _transactionStorage = transactionStorage;
+            _receiptStorage = receiptStorage;
+            _signer = signer;
+            _peerThreshold = peerThreshold;
+        }
+
+        public Transaction[] PendingTransactions => _pendingTransactions.Values.ToArray();
+        public TransactionReceipt GetReceipt(Keccak hash) => _receiptStorage.Get(hash);
+
+        public void AddFilter<T>(T filter) where T : ITransactionFilter
+            => _filters.TryAdd(filter.GetType(), filter);
+
+        public void DeleteFilter(Type type) => _filters.TryRemove(type, out _);
+
+        public void AddPeer(ISynchronizationPeer peer)
+        {
+            if (!_peers.TryAdd(peer.NodeId.PublicKey, peer))
+            {
+                return;
+            }
+
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Added peer: {peer.ClientId}");
+            }
+        }
+
+        public void DeletePeer(NodeId nodeId)
+        {
+            if (!_peers.TryRemove(nodeId.PublicKey, out _))
+            {
+                return;
+            }
+
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Removed peer: {nodeId}");
+            }
+        }
+
+        public void AddTransaction(Transaction transaction, UInt256 blockNumber)
+        {
+            if (_pendingTransactions.ContainsKey(transaction.Hash))
+            {
+                return;
+            }
+
+            var recoveredAddress = _signer.RecoverAddress(transaction, blockNumber);
+            if (recoveredAddress != transaction.SenderAddress)
+            {
+                throw new InvalidOperationException("Invalid signature");
+            }
+
+            _pendingTransactions.TryAdd(transaction.Hash, transaction);
+            NewPending?.Invoke(this, new TransactionEventArgs(transaction));
+            NotifyPeers(SelectPeers(GetAvailablePeers(transaction)), transaction);
+            FilterAndStoreTransaction(transaction, blockNumber);
+        }
+
+        private void FilterAndStoreTransaction(Transaction transaction, UInt256 blockNumber)
+        {
+            var filters = _filters.Values.ToArray();
+            for (var i = 0; i < filters.Length; i++)
+            {
+                var filter = filters[i];
+                if (!filter.CanAdd(transaction))
+                {
+                    return;
+                }
+            }
+
+            _transactionStorage.Add(transaction, blockNumber);
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Added transaction: {transaction.Hash}");
+            }
+        }
+
+        public void DeleteTransaction(Keccak hash)
+        {
```

Comment:
what is he use case for delete filter?

Reply:
TxPool keeps track of pending transactions internally, while filters add (or not) these transactions to the given storage (if provided). The question is - should a filter also decide, whether a transaction can be removed from storage?